examples/pytorch/question-answering/sparseml_utils.py

```diff
@@ -0,0 +1,205 @@
+
+    :param nm_prune_config: recipe for model sparsification
+    :param args, kwargs: arguments passed into parent class
+    """
+
+    def __init__(self, nm_prune_config, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.nm_prune_config = nm_prune_config
+
+    def create_optimizer(self):
+        """
+        Create optimizer customized using SparseML
+        """
+        super().create_optimizer()
+        steps_per_epoch = math.ceil(
+            len(self.train_dataset) / (self.args.per_device_train_batch_size * self.args._n_gpu)
+        )
+        manager = ScheduledModifierManager.from_yaml(self.nm_prune_config)
+        self.args.num_train_epochs = float(manager.max_epochs)
+        if hasattr(self, "scaler"):
+            manager.initialize(self.model, epoch=0.0)
+            self.scaler = manager.modify(
+                self.model, self.optimizer, steps_per_epoch=steps_per_epoch, wrap_optim=self.scaler
+            )
+        else:
+            self.optimizer = ScheduledOptimizer(
+                self.optimizer, self.model, manager, steps_per_epoch=steps_per_epoch, loggers=None
+            )
+
+
+class SparseMLDistillQATrainer(SparseMLQATrainer):
+    """
+    Question Answering trainer using distilation with customized optimizer using SparseML
+
+    :param nm_prune_config: recipe for model sparsification
+    :param teacher: teacher model
+    :param distill_hardness: weight of the teacher loss
+    :param temperature: temperature used for loss
+    :param args, kwargs: arguments passed into parent class
+    """
+
+    def __init__(self, nm_prune_config, teacher=None, distill_hardness=0.5, temperature=2.0, *args, **kwargs):
+        super().__init__(nm_prune_config, *args, **kwargs)
+        self.teacher = teacher
+        self.distill_hardness = distill_hardness
+        self.temperature = temperature
+        self.criterion = torch.nn.CrossEntropyLoss()
+
+    def compute_loss(self, model, inputs, return_outputs=False):
+        """
+        Computing loss using teacher/student distillation
+        """
+        outputs = model(**inputs)
+        loss = outputs["loss"]
+        if self.teacher is not None:
+            input_device = inputs["input_ids"].device
+            self.teacher = self.teacher.to(input_device)
+            start_logits_student = outputs["start_logits"]
+            end_logits_student = outputs["end_logits"]
+            start_logits_label = inputs["start_positions"]
+            end_logits_label = inputs["start_positions"]
+            with torch.no_grad():
+                teacher_output = self.teacher(
+                    input_ids=inputs["input_ids"],
+                    token_type_ids=inputs["token_type_ids"],
+                    attention_mask=inputs["attention_mask"],
+                )
+            start_logits_teacher = teacher_output["start_logits"]
+            end_logits_teacher = teacher_output["end_logits"]
+            loss_start = (
```

Comment:
we have a function in SparseML for distillation, would like to update that to match this and use that

Reply:
I'll need to update this part later